consensus/istanbul/backend/engine.go

```diff
@@ -921,10 +921,69 @@ func (sb *backend) snapshot(chain consensus.ChainReader, number uint64, hash com
 		logger.Trace("Stored voting snapshot to disk", "number", snap.Number, "hash", snap.Hash)
 	}
 
+	sb.regen(chain, headers)
+
 	sb.recents.Add(snap.Hash, snap)
 	return snap, err
 }
 
+// regen commits snapshot data to database
+// regen is triggered if there is any checkpoint block in the `headers`.
+// For each checkpoint block, this function verifies the existence of its snapshot in DB and stores one if missing.
+/*
+ Triggered:
+ |   ^                          ^                          ^                          ^  ...|
+     SI                 SI*(last snapshot)                 SI                         SI
+       			   | header1, .. headerN |
+ Not triggered: (Guaranteed SI* was committed before )
+ |   ^                          ^                          ^                          ^  ...|
+     SI                 SI*(last snapshot)                 SI                         SI
+	                            | header1, .. headerN |
+*/
+func (sb *backend) regen(chain consensus.ChainReader, headers []*types.Header) {
+	if !sb.isRestoringSnapshots.Load() && len(headers) > 1 {
```

Comment:
Minor comments related to the style. 
1. Reducing the call depth 
```
if sb.isRestoringSnapshots.Load() || len(headers) < 1 {  
    return 
}
```
BTW, is it okay to pass `len(headers) == 1` case in the original code? 

2. More safe operation for the atomic value change
`sb.isRestoringSnapshots.Load()` -> `sb.isRestoringSnapshots.CompareAndSwap(false, true)`

Reply:
Thanks for the good suggestions.

First, I'll take your first suggestion. Once the suggestion is updated, the residual task will be as follows:
Whether to change `Store` to `CompareAndSwap`.
```
sb.isRestoringSnapshots.Store(true) // sb.isRestoringSnapshots.CompareAndSwap(false, true)
		defer func() {
			sb.isRestoringSnapshots.Store(false) // sb.isRestoringSnapshots.CompareAndSwap(true, false)
		}()
```
I think the existing code seems more intuitive and reliable.

Additionally, I couldn't understand what this implies.

> BTW, is it okay to pass len(headers) == 1 case in the original code?

Could you give a detail? BTW, I installed a guard from proceeding when the length of header is one because it was already handled before `regen` handler is called.

Reply:
I noticed before that load-store can have race conditions, but ignored because it's practically impossible to trigger.
But now that it's mentioned, safer design would be better.

compare-and-swap lock goes like this (it doesn't need load or store):

```go
lock := uint32(0)
if sb.isRestoringSnapshots.CompareAndSwap(false, true) {
  // enter critical section
}
```

Note that previous atomic load-store can have race conditions if two threads are running in parallel.
```
T1 - load and check
T2 - load and check
T1 - store
T2 - store
both threads enter critical section
```

Reply:
Possibly make sense. Thank you. Before making changes, let me ask one concern from your comment. Once the critical section executed, restoring `true` to `false` seems still required. Doesnt' it?

Reply:
Yep, you're right

Reply:
@ian0371 Updated. PTAL.
@aidan-kwon, I kept `len(headers) <= 1`. PTAL.